In [1]:
!pip install langchain langchain-community pypdf2 python-dotenv openai tiktoken

  Using cached langchain-1.3.4-py3-none-any.whl.metadata (5.8 kB)
  Using cached langgraph-1.2.4-py3-none-any.whl.metadata (8.0 kB)
  Using cached langchain_protocol-0.0.16-py3-none-any.whl.metadata (2.4 kB)
  Using cached langsmith-0.8.9-py3-none-any.whl.metadata (15 kB)
  Using cached langgraph_checkpoint-4.1.1-py3-none-any.whl.metadata (5.2 kB)
  Using cached langgraph_prebuilt-1.1.0-py3-none-any.whl.metadata (5.2 kB)
  Using cached langgraph_sdk-0.4.2-py3-none-any.whl.metadata (3.6 kB)
Using cached langchain-1.3.4-py3-none-any.whl (125 kB)
   ---------------------------------------- 0.0/549.1 kB ? eta -:--:--
   ---------------------------------------- 549.1/549.1 kB 4.4 MB/s  0:00:00
Using cached langgraph-1.2.4-py3-none-any.whl (245 kB)
Using cached langgraph_checkpoint-4.1.1-py3-none-any.whl (56 kB)
Using cached langgraph_prebuilt-1.1.0-py3-none-any.whl (41 kB)
Using cached langgraph_sdk-0.4.2-py3-none-any.whl (160 kB)
Using cached langsmith-0.8.9-py3-none-any.whl (403 kB)
   --

In [3]:
from langchain_text_splitters import (
    CharacterTextSplitter,
    RecursiveCharacterTextSplitter,
    TokenTextSplitter
)

import PyPDF2

In [6]:
import PyPDF2

pdf_path = "data\hleg_ethics_guidelines_for_trustworthy_ai-en.pdf"

pdf_text = ""

with open(pdf_path, "rb") as file:
    reader = PyPDF2.PdfReader(file)

    for page in reader.pages:
        text = page.extract_text()
        if text:
            pdf_text += text + "\n"

print(f"PDF loaded: {len(pdf_text)} characters")
print(pdf_text[:500])


PDF loaded: 158079 characters
 
  
 
 
 
INDEPENDENT  
HIGH-LEVEL EXPERT GROUP ON  
ARTIFICIAL INTELLIGENCE  
SET UP BY THE EUROPEAN COMMISSION  
 
 
 
 
 
 
ETHICS GUIDELINES  
FOR TRUSTWORTHY AI 
 
 
 
  
 
ETHICS GUIDELINES  FOR  TRUSTWORTHY  AI 
 
High -Level Expert Group on Artificial Intelligence  
 
 
 
 
 
 
 
 
This document was written by the High -Level Expert Group on AI (AI HLEG) . The members of the AI HLEG 
named in this document support the overall framework for Trustworthy  AI put forward in these Guidelines


In [7]:
print(f"Total PDF characters: {len(pdf_text):,}")

pdf_words = len(pdf_text.split())

print(f"Approximate words: {pdf_words:,}")

Total PDF characters: 158,079
Approximate words: 24,153


In [15]:
start = 22000
end = 27000

pdf_sample = pdf_text[start:end]

print(pdf_sample)

 be able  to 
participate by indicating their interest through the European AI Alliance.  
 
B. A FRAMEWORK FOR TRUSTWORTHY  AI 
These  Guidelines articulate a framework for achiev ing Trustworthy  AI based  on fundamental rights as enshrined in 
the Charter of Fundamental Rights of the European Union (EU Charter ), and in relevant international human rights  
law. Below, we briefly touch upon Trustworthy  AI’s three components.  
Lawful AI  
AI systems do not operate in a lawless world. A number of legally binding rules at European, national and 
international level already apply or are relevant to the development , deployment  and use of AI systems today. Legal 
sources include , but are not limit ed to : EU p rimary law (the Treaties of the  European Union and its Charter of 
Fundamental Rights ), EU s econdary law (such as the General Data Protection Regulation, the Product Liability 
Directive , the Regulati on on the Free Flow of Non -Personal D ata, anti-discrimination Directive

In [16]:
fixed_splitter = CharacterTextSplitter(
    chunk_size=1000,
    chunk_overlap=100,
    separator="\n"
)

pdf_chunks_fixed = fixed_splitter.split_text(pdf_text)

print(f"Number of chunks: {len(pdf_chunks_fixed)}")

for i in range(3):
    print(f"\n--- Chunk {i+1} ---\n")
    print(pdf_chunks_fixed[i][:300])

Number of chunks: 171

--- Chunk 1 ---

INDEPENDENT  
HIGH-LEVEL EXPERT GROUP ON  
ARTIFICIAL INTELLIGENCE  
SET UP BY THE EUROPEAN COMMISSION  
 
 
 
 
 
 
ETHICS GUIDELINES  
FOR TRUSTWORTHY AI 
 
 
 
  
 
ETHICS GUIDELINES  FOR  TRUSTWORTHY  AI 
 
High -Level Expert Group on Artificial Intelligence  
 
 
 
 
 
 
 
 
This document was w

--- Chunk 2 ---

The AI HLEG is an independent expert group that was set up by the European Commission in June 2018.  
 
Contact  Nathalie Smuha - AI HLEG Coordinator  
E-mail  CNECT -HLG-AI@ec.europa.eu  
 
European Commission  
B-1049 Brussels  
 
Document made public on 8 April 2019.  
 
A first draft of this doc

--- Chunk 3 ---

Guidelines, the views expressed in this document  reflect the opinion of the AI HLEG  and may not in any circumstances be 
regarded as reflecting  an official position of the European Commission.  
 
More information on the High -Level Exp ert Group on Artificial Intelligence is available online  (h


### Fixed-size PDF chunking observations

Using CharacterTextSplitter with chunk_size=1000 and chunk_overlap=100 created 171 chunks. The chunks are mostly readable, but several boundaries cut sentences or context awkwardly. This shows that fixed-size chunking is simple and predictable, but it does not reliably preserve sentence or section boundaries in the PDF.

In [17]:
recursive_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000,
    chunk_overlap=200,
    length_function=len,
    separators=["\n\n", "\n", ". ", " ", ""]
)

pdf_chunks_recursive = recursive_splitter.split_text(pdf_text)

print(f"Number of chunks: {len(pdf_chunks_recursive)}")

for i in range(3):
    print(f"\n--- Chunk {i+1} ---\n")
    print(pdf_chunks_recursive[i][:300])

Number of chunks: 196

--- Chunk 1 ---

INDEPENDENT  
HIGH-LEVEL EXPERT GROUP ON  
ARTIFICIAL INTELLIGENCE  
SET UP BY THE EUROPEAN COMMISSION  
 
 
 
 
 
 
ETHICS GUIDELINES  
FOR TRUSTWORTHY AI 
 
 
 
  
 
ETHICS GUIDELINES  FOR  TRUSTWORTHY  AI 
 
High -Level Expert Group on Artificial Intelligence  
 
 
 
 
 
 
 
 
This document was w

--- Chunk 2 ---

feedback gathered through the piloting phase, will be presented to the Eu ropean Commission in early 2020.  
 
 
The AI HLEG is an independent expert group that was set up by the European Commission in June 2018.  
 
Contact  Nathalie Smuha - AI HLEG Coordinator  
E-mail  CNECT -HLG-AI@ec.europa.eu 

--- Chunk 3 ---

might be made of the following information. The contents of this working document  are the sole responsibility of the High -
Level Expert Group on Artificial Intelligence (AI HLEG). Although Commission staff  facilitated  the preparation of the 
Guidelines, the views expressed in this document  refl


In [18]:
print(pdf_chunks_fixed[50][:500])

print("\n" + "="*50 + "\n")

print(pdf_chunks_recursive[50][:500])

and ensuring that individuals and groups are free from unfair bias, discrimination and stigmatisation. If unfair biases 
can be avoided, AI systems could even increase societal fair ness. E qual opportunity in terms of access to education, 
goods, services  and technology should also be fostered . Moreover, the use of AI systems should never lead to 
people  being deceived or  unjustifiably impaired in their freedom of choice. Additionally, fairness implies that AI 
practitioners should respect 


2.2 Ethical Principles in the Context of AI  Systems22 
Many public, private, and civil organizations have drawn inspiration from fundamental rights to produce ethical 
frameworks for AI  systems .23 In the EU, the European Group on Ethics in Science and New Technologies (“EGE”) 
proposed a set of 9 basic principles, based on the fundamental values laid down in the EU Treaties and Charter.24 
We build  further on this work, recognising most of the principles hitherto propounded by various gro

### Recursive PDF chunking observations

RecursiveCharacterTextSplitter created 196 chunks. Compared with fixed-size chunking, the chunks were more likely to start at meaningful section or paragraph boundaries, such as section headings. However, some chunks were still quite short or incomplete because the extracted PDF text contains formatting artefacts and irregular line breaks. Overall, recursive chunking preserved document structure better than fixed-size chunking, but the result still depends heavily on the quality of PDF text extraction.

In [19]:
token_splitter = TokenTextSplitter(
    chunk_size=250,
    chunk_overlap=50
)

pdf_chunks_token = token_splitter.split_text(pdf_text)

print(f"Number of chunks: {len(pdf_chunks_token)}")

for i in range(3):
    print(f"\n--- Chunk {i+1} ---\n")
    print(pdf_chunks_token[i][:300])

Number of chunks: 191

--- Chunk 1 ---

 
  
 
 
 
INDEPENDENT  
HIGH-LEVEL EXPERT GROUP ON  
ARTIFICIAL INTELLIGENCE  
SET UP BY THE EUROPEAN COMMISSION  
 
 
 
 
 
 
ETHICS GUIDELINES  
FOR TRUSTWORTHY AI 
 
 
 
  
 
ETHICS GUIDELINES  FOR  TRUSTWORTHY  AI 
 
High -Level Expert Group on Artificial Intelligence  
 
 
 
 
 
 
 
 
This doc

--- Chunk 2 ---

 with every single statement in the document .  
 
The Trustworthy  AI assessment  list presented in Chapter III of this document will undergo a piloting phase by 
stakeholders to gather practical feedback. A revised version  of the assessment  list, taking into account the 
feedback gathered throug

--- Chunk 3 ---

 subject to an open consultation which 
generated feedback from more than 500 contributors. We wish to explicitly and warmly tha nk all those who contributed 
their feedback on the document's first draft, which was considered in the preparation of this revised version.  
 
Neither the European Commi


In [20]:
print(pdf_chunks_token[50][:500])

 imposing a hierarchy, we list the principles here below in manner that  mirrors the order of 
appearance of the fundamental rights upon which they are based  in the EU Charter.25    
                                                           
21  For a description of the term as used throughout this document,  see the Glossary.  
22  These principles also apply to the development, deployment and use of other technologies, and hence are not specific to AI 
systems. In what follows, we have aimed


### Token-based PDF chunking observations

Token-based chunking produced chunks of a more predictable size for use with large language models. However, it did not consistently preserve document structure or section boundaries. Compared with recursive chunking, the resulting chunks appeared less semantically coherent, although they remained more suitable for managing LLM context limits than fixed-size character chunking.